### PR TITLE
`azurerm_virtual_network` - replace `address_prefix` with `address_prefixes` in 4.0

### PR DESCRIPTION
**Pull request by stephybun**
## Community Note

* Please vote on this PR by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original PR to help the community and maintainers prioritize for review
* Please do not leave comments along the lines of "+1", "me too" or "any updates", they generate extra noise for PR followers and do not help prioritize for review

## Description

Related to #3917

## Testing 

Tests in 4.0:
```
--- PASS: TestAccVirtualNetwork_basic (75.65s)
--- PASS: TestAccVirtualNetwork_complete (88.93s)
--- PASS: TestAccVirtualNetwork_disappears (114.28s)
--- PASS: TestAccVirtualNetwork_basicUpdated (120.78s)
--- PASS: TestAccVirtualNetwork_requiresImport (122.83s)
--- PASS: TestAccVirtualNetwork_ddosProtectionPlan (128.28s)
--- PASS: TestAccVirtualNetwork_deleteSubnet (138.33s)
--- PASS: TestAccVirtualNetwork_subnetRouteTable (169.76s)
--- PASS: TestAccVirtualNetwork_withTags (172.84s)
--- PASS: TestAccVirtualNetwork_bgpCommunity (190.74s)
--- PASS: TestAccVirtualNetwork_edgeZone (244.48s)
--- PASS: TestAccVirtualNetwork_subnet (247.60s)
--- PASS: TestAccVirtualNetwork_updateFlowTimeoutInMinutes (257.57s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/network       260.613s

```

Tests in 3.0:
```
--- PASS: TestAccVirtualNetwork_disappears (81.34s)
--- PASS: TestAccVirtualNetwork_requiresImport (105.60s)
--- PASS: TestAccVirtualNetwork_withTags (119.88s)
--- PASS: TestAccVirtualNetwork_basic (128.33s)
--- PASS: TestAccVirtualNetwork_complete (133.41s)
--- PASS: TestAccVirtualNetwork_ddosProtectionPlan (145.43s)
--- PASS: TestAccVirtualNetwork_basicUpdated (151.87s)
--- PASS: TestAccVirtualNetwork_edgeZone (181.18s)
--- PASS: TestAccVirtualNetwork_deleteSubnet (182.49s)
--- PASS: TestAccVirtualNetwork_updateFlowTimeoutInMinutes (204.86s)
--- PASS: TestAccVirtualNetwork_bgpCommunity (206.07s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/network       209.177s

```

Manual upgrade from v3.114.0 to 4.0:
```
Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # azurerm_virtual_network.test will be updated in-place
  ~ resource "azurerm_virtual_network" "test" {
        id                      = "/subscriptions/1a6092a6-137e-4025-9a7c-ef77f76f2c02/resourceGroups/bcctest-sw/providers/Microsoft.Network/virtualNetworks/acctestvirtnetsw"
        name                    = "acctestvirtnetsw"
        tags                    = {}
        # (6 unchanged attributes hidden)

      - subnet {
          - address_prefixes                              = [
              - "10.0.1.0/24",
            ] -> null
          - default_outbound_access_enabled               = false -> null
          - id                                            = "/subscriptions/1a6092a6-137e-4025-9a7c-ef77f76f2c02/resourceGroups/bcctest-sw/providers/Microsoft.Network/virtualNetworks/acctestvirtnetsw/subnets/subnet1" -> null
          - name                                          = "subnet1" -> null
          - private_endpoint_network_policies             = "Disabled" -> null
          - private_link_service_network_policies_enabled = true -> null
          - service_endpoint_policy_ids                   = [] -> null
          - service_endpoints                             = [] -> null
        }
      - subnet {
          - address_prefixes                              = [
              - "10.10.1.0/24",
            ] -> null
          - default_outbound_access_enabled               = false -> null
          - id                                            = "/subscriptions/1a6092a6-137e-4025-9a7c-ef77f76f2c02/resourceGroups/bcctest-sw/providers/Microsoft.Network/virtualNetworks/acctestvirtnetsw/subnets/subnet2" -> null
          - name                                          = "subnet2" -> null
          - private_endpoint_network_policies             = "Disabled" -> null
          - private_link_service_network_policies_enabled = true -> null
          - service_endpoint_policy_ids                   = [] -> null
          - service_endpoints                             = [] -> null
        }
      + subnet {
          + address_prefixes                              = [
              + "10.0.1.0/24",
            ]
          + default_outbound_access_enabled               = true
          + id                                            = (known after apply)
          + name                                          = "subnet1"
          + private_endpoint_network_policies             = "Disabled"
          + private_link_service_network_policies_enabled = true
          + service_endpoint_policy_ids                   = []
          + service_endpoints                             = []
        }
      + subnet {
          + address_prefixes                              = [
              + "10.10.1.0/24",
            ]
          + default_outbound_access_enabled               = true
          + id                                            = (known after apply)
          + name                                          = "subnet2"
          + private_endpoint_network_policies             = "Disabled"
          + private_link_service_network_policies_enabled = true
          + service_endpoint_policy_ids                   = []
          + service_endpoints                             = []
        }

        # (1 unchanged block hidden)
    }

Plan: 0 to add, 1 to change, 0 to destroy.
```

### CODE REVIEW

File: internal/services/network/virtual_network_resource.go [hunk 1]
```diff
@@ -753,9 +774,22 @@ func expandVirtualNetworkProperties(ctx context.Context, client virtualnetworks.
...
+				for _, prefix := range subnet["address_prefixes"].([]interface{}) {
+					addressPrefixes = append(addressPrefixes, prefix.(string))
+				}
+
+				if len(addressPrefixes) == 1 {
+					subnetObj.Properties.AddressPrefix = pointer.To(addressPrefixes[0])
+				} else {
+					subnetObj.Properties.AddressPrefixes = pointer.To(addressPrefixes)
+				}
+
```
Comment (jackofallops) [inline internal/services/network/virtual_network_resource.go:792]: Do we need to `nil` out anything here to account for change from 1 prefix to 2+?

**Review by jackofallops — APPROVED** [review]
Thanks @stephybun - just a couple comments to take a look at, but otherwise LGTM 👍

File: internal/services/network/virtual_network_resource.go (hunk 1 above)
Comment (stephybun) [reply internal/services/network/virtual_network_resource.go:792]: This part is only called in the create, so there shouldn't be anything to `nil` here

File: internal/services/network/virtual_network_resource.go [hunk 2]
```diff
@@ -684,9 +688,26 @@ func expandVirtualNetworkSubnets(ctx context.Context, client virtualnetworks.Vir
...
+			if len(addressPrefixes) == 1 {
+				subnetObj.Properties.AddressPrefix = pointer.To(addressPrefixes[0])
+				subnetObj.Properties.AddressPrefixes = nil
+			} else {
+				subnetObj.Properties.AddressPrefixes = pointer.To(addressPrefixes)
+				subnetObj.Properties.AddressPrefix = nil
+			}
+
+			subnetObj.Properties.AddressPrefixes = pointer.To(addressPrefixes)
```
[the remark this answers is not in the document]
Comment (stephybun) [reply internal/services/network/virtual_network_resource.go]: Turns out line 707 should be removed anyway!

**Review by manicminer — APPROVED** [review]

**Pull request merged**